BlockchainSdk/Blockchains/Cardano/CardanoWalletManager.swift

```diff
@@ -102,36 +104,131 @@ extension CardanoWalletManager: TransactionSender {
+            return .cardanoWillBeSendAlongToken(amount: amount)
+
+        } catch {
+            print("CardanoWalletManager", #function, "catch error: \(error)")
+            return nil
+        }
+    }
+}
+
+// MARK: - CardanoTransferRestrictable
+
+extension CardanoWalletManager: CardanoTransferRestrictable {
+    func validateCardanoTransfer(amount: Amount, fee: Amount) throws {
+        switch amount.type {
+        case .coin:
+            let hasTokensWithBalance = wallet.amounts.contains { amountType, amount in
+                amountType.isToken && amount.value > 0
+            }
+
+            guard hasTokensWithBalance else {
+                // Skip this checking. Dust checking will be after
+                return
+            }
+            
+            try validateCardanoCoinWithdrawal(amount: amount, fee: fee)
+        case .token:
+            try validateCardanoTokenWithdrawal(amount: amount, fee: fee)
+        case .reserve:
+            throw BlockchainSdkError.notImplemented
+        }
+    }
+
+    private func validateCardanoCoinWithdrawal(amount: Amount, fee: Amount) throws {
+        assert(!amount.type.isToken, "Only coin validation")
+
+        guard let adaBalance = wallet.amounts[.coin]?.value else {
+            throw ValidationError.balanceNotFound
+        }
+
+        let minChange = try minChange(amount: amount)
+        var change = adaBalance - amount.value
+
+        if amount.type == fee.type {
+            change -= fee.value
+        }
+
+        if change < minChange.value {
+            throw ValidationError.cardanoHasTokens(minimumAmount: minChange)
+        }
+    }
+
+    private func validateCardanoTokenWithdrawal(amount: Amount, fee: Amount) throws {
+        assert(amount.type.isToken, "Only token validation")
+
+        guard var adaBalance = wallet.amounts[.coin]?.value else {
+            throw ValidationError.balanceNotFound
+        }
+
+        guard let tokenBalance = wallet.amounts[amount.type]?.value else {
+            throw ValidationError.balanceNotFound
+        }
+
+        // the fee will be spend in any case
+        adaBalance -= fee.value
+        
+        // 1. Check if there is enough ADA to send the token
+        let minAdaValue = try transactionBuilder.buildCardanoSpendingAdaValue(amount: amount, fee: fee)
+        let minAdaToSendDecimal = Decimal(minAdaValue) / wallet.blockchain.decimalValue
+
+        // Not enough balance to send token
+        if minAdaToSendDecimal > adaBalance {
+            throw ValidationError.cardanoInsufficientBalanceToSendToken
+        }
+
+        // 2. Check if there is enough ADA to get a change with after transaction
+        let minChange = try minChange(amount: amount)
+        let change = adaBalance - minAdaToSendDecimal
+
+        // If there not enough ada balance to change
+        guard change > 0, change < minChange.value else {
+            return
+        }
+
+        throw ValidationError.cardanoInsufficientBalanceToSendToken
```

Comment:
Согласен с Денисом, что странно получается. Давай лучше вот так
```suggestion
        if 0 < change, change < minChange.value {
            throw ValidationError.cardanoInsufficientBalanceToSendToken
        }
```

кстати, может проверка должна быть 
``` swift
    if change < 0 || change < minChange.value {
        throw ValidationError.cardanoInsufficientBalanceToSendToken
    }
```

Reply:
Обновил на `if`. Не тут правильная проверка, что сдача "есть" и что она "больше чем минимум"